core/src/main/scala/kafka/coordinator/GroupMetadataManager.scala

```diff
@@ -0,0 +1,919 @@
+          error("Failed to mark group %s as deleted in %s.".format(groupId, appendPartition), t)
+          // ignore and continue
+      }
+    }
+  }
+
+  def storeGroup(group: GroupMetadata,
+                 groupAssignment: Map[String, Array[Byte]]) {
+    // construct the message to append
+    val message = new Message(
+      key = GroupMetadataManager.groupMetadataKey(group.groupId),
+      bytes = GroupMetadataManager.groupMetadataValue(group, groupAssignment)
+    )
+
+    val groupMetadataPartition = TopicAndPartition(GroupCoordinator.GroupMetadataTopicName, partitionFor(group.groupId))
+
+    val groupMetadataMessageSet = Map(groupMetadataPartition ->
+      new ByteBufferMessageSet(config.offsetsTopicCompressionCodec, message))
+
+    // set the callback function to insert offsets into cache after log append completed
+    def putCacheCallback(responseStatus: Map[TopicAndPartition, ProducerResponseStatus]) {
+      // the append response should only contain the topics partition
+      if (responseStatus.size != 1 || ! responseStatus.contains(groupMetadataPartition))
+        throw new IllegalStateException("Append status %s should only have one partition %s"
+          .format(responseStatus, groupMetadataPartition))
+
+      // construct the commit response status and insert
+      // the offset and metadata to cache if the append status has no error
+      val status = responseStatus(groupMetadataPartition)
+
+      var responseCode = Errors.NONE.code
+      if (status.error != ErrorMapping.NoError) {
+        debug("Metadata from group %s with generation %d failed when appending to log due to %s"
+          .format(group.groupId, group.generationId, ErrorMapping.exceptionNameFor(status.error)))
+
+        // transform the log append error code to the corresponding the commit status error code
+        responseCode = if (status.error == ErrorMapping.UnknownTopicOrPartitionCode) {
+          Errors.GROUP_COORDINATOR_NOT_AVAILABLE.code
+        } else if (status.error == ErrorMapping.NotLeaderForPartitionCode) {
+          Errors.NOT_COORDINATOR_FOR_GROUP.code
+        } else if (status.error == ErrorMapping.MessageSizeTooLargeCode
+          || status.error == ErrorMapping.MessageSetSizeTooLargeCode
+          || status.error == ErrorMapping.InvalidFetchSizeCode) {
+
+          error("Appending metadata message for group %s generation %d failed due to %s, returning UNKNOWN error code to the client"
+            .format(group.groupId, group.generationId, ErrorMapping.exceptionNameFor(status.error)))
+
+          Errors.UNKNOWN.code
+        } else {
+
+          error("Appending metadata message for group %s generation %d failed due to unexpected error: %s"
+            .format(group.groupId, group.generationId, status.error))
+
+          status.error
+        }
+      }
+
+      for (member <- group.allMembers) {
+        member.assignment = groupAssignment.getOrElse(member.memberId, Array.empty[Byte])
```

Comment:
We've already transitioned to Stable, but the assignments will not be set until the callback is invoked. Is it possible for a follower SyncGroup to arrive before the callback is invoked?